Runtime: avoid an assertion in lldb

compnerd opened the pull request:
When lldb is running the testsuite, it can result in this assertion triggering.
However, since we now have an absolute pointer, we do not do any pointer
tagging, so remove the assertion.

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd commented:
CC: @rjmccall @adrian-prantl @llvm-beanz @jimingham

adrian-prantl commented:
@swift-ci test

jckarter commented:
The assertion is valid—`Description` doesn't exist on a non-Swift class and would read junk memory. I wouldn't expect your change to cause the assertion to fire—have we looked into why?

compnerd commented:
@jckarter during the expression evaluation, lldb is trying to determine the dynamic type of the expression.  This results in it reading the metadata for the type, which will in turn attempt to read the nominal type descriptor.  The metadata indicates that the type is a class, so we will attempt to query the description in `readAddressOfNominalTypeDescriptor`.  I guess the question then is, why does the runtime think that the returned type (in this particular case, a `NSString`) is a class?  Because `NSString` is bridged into Swift, the Description needs to be present.

Is my understanding flawed?

jckarter commented:
lldb ought to be checking whether the class is a Swift class before trying to access the nominal type descriptor, since ObjC classes don't have one, and don't allocate storage for one.

compnerd commented:
Hmm, I don't know how to determine whether we are interacting with a swift object.  Any pointers on that?  @jimingham?

jckarter commented:
You can do the same test the assertion is doing, `isTypeMetadata()`.

jckarter commented:
(If lldb is looking at the bits and not using our C++ interfaces, I believe that checks a reserved bit in the ObjC matter's rodata pointer. cc @gparker42)

compnerd commented:
@jckarter hmm, we could expose that through the remote AST perhaps?

In `SwiftLanguageRuntime.cpp` `SwiftLanguageRuntime::MetadataPromise::FulfillTypePromise` will do a `m_remote_aast->getTypeForRemoteTypeMetadata` which will trigger the invalid access.

jckarter commented:
Calling @slavapestov. Technically defined-in-ObjC class objects aren't Swift type metadata, so I'm not sure we should be aiming RemoteAST at one to begin with, though perhaps it could fail more gracefully when you try.

compnerd commented:
@jckarter yeah, it definitely shouldn't fail miserably like that.  Plus, since this code is technically part of the runtime, I think that the `NDEBUG` (i.e. NoAsserts) behavior needs to be correct too.  Jim is of the opinion that lldb shouldn't be required to determine what the origin of the object is (that is, if it is in a swift module, it should be able to treat it as a swift object).  That does certainly simplify implementation of the debugger.

jimingham commented:
That's right.  One use of this interface, for instance, is when the debugger stops in a function where a variable is passed as a generic type which is statically known only by protocol conformance.  We're want to ask the RemoteAST what it actually is.  Before we get an answer, we have no idea what the object is.  

We've been trying to get the debugger away from having to peek at bits in the objects, and instead ask swift what they are.  That seems a much more supportable design.

jckarter commented:
Sure, I agree that we should provide a proper interface for this. As is, though, this would end up handing off garbage after the `NSString` class object as if it were a nominal type descriptor, which is something that doesn't even exist for purely-ObjC classes. If we can modify RemoteAST to check for the is-Swift bit, and there's something else we use in place of nominal type descriptors for ObjC classes, then it would make sense to me to do that.

compnerd commented:
@jckarter another option is to have `readTypeFromMetadata` check if the type is swift based.  Would you be okay with that?

lhames commented:
Hi @compnerd,

This is still breaking the LLDB test suite: https://ci.swift.org/job/oss-lldb-incremental-osx

What's the timeline for the proper fix?
If this can't be fixed in the next hour or so, can you please revert the offending patch for now?

compnerd committed:
RemoteAST: only swift has nominal type descriptors

When the RemoteAST is used to access objects from LLDB, the object that
we are interacting with may not be a Swift object but rather an ObjC
object that has been bridged into swift.  In such a case, the object
does not have a nominal type descriptor associated with it.  Ensure that
the type is a valid Swift type before reading the nominal type
descriptor from the metadata.  This avoids an assertion when evaluating
expressions in LLDB.

jckarter commented:
@compnerd Sure, if there's reasonable fallback behavior for an ObjC class.

compnerd commented:
@lhames was running through the lldb test suite to make sure that this fixed all the cases :-)

compnerd commented:
@swift-ci please test

lhames commented:
@compnerd Thanks for that. So we're just waiting on these CI tests now before this can go in?

compnerd commented:
@lhames yeah :-)

lhames commented:
Thanks very much! :)
